### PR TITLE
feat(errors): generate sarif report on successful execution

### PR DESCRIPTION
**Pull request by royalpinto007**
- Add SARIF report generation for successful executions, ensuring consistent error reporting format. This change outputs a minimal SARIF structure with `executionSuccessful: true` and an empty results array, providing uniformity in SARIF output across all executions.
- Use regex for version.

### CODE REVIEW

File: compiler/src/dmd/sarif.d [hunk 1]
```diff
@@ -165,44 +165,16 @@ Params:
...
+  executionSuccessful = `true` if execution succeeded; includes an empty `results` array in the SARIF report.
 
 Throws:
   This function is marked as `nothrow` and does not throw exceptions.
 
 See_Also:
   $(LINK2 https://schemastore.azurewebsites.net/schemas/json/sarif-2.1.0.json, SARIF 2.1.0 schema)
 */
-void generateSarifReport(const ref SourceLoc loc, const(char)* format, va_list ap, ErrorKind kind) nothrow
+void generateSarifReport(const ref SourceLoc loc, const(char)* format = null, va_list ap = null, ErrorKind kind = ErrorKind.message, bool executionSuccessful = true) nothrow
```
Comment (thewilsonator) [inline compiler/src/dmd/sarif.d:176]: why the default parameters? also `executionSuccessful` is `false` for most call site, why default it into true?

File: compiler/src/dmd/sarif.d [hunk 2]
```diff
@@ -257,67 +229,106 @@ void generateSarifReport(const ref SourceLoc loc, const(char)* format, va_list a
...
-    ob.writestring(`"startLine": `);
-    ob.printf(`%d,`, loc.linnum);
-    ob.writestringln(``);
-    ob.writestring(`"startColumn": `);
-    ob.printf(`%d`, loc.charnum);
-    ob.writestringln(``);
-    ob.level -= 1;
-    ob.writestringln(`}`);
+    // Error information if execution was unsuccessful
+    if (!executionSuccessful && format && ap) {
```
[the remark this answers is not in the document]
Comment (thewilsonator) [reply compiler/src/dmd/sarif.d:239]: also invert this condition so that the very small `else` case comes first.

File: compiler/src/dmd/main.d [hunk 3]
```diff
@@ -172,6 +173,14 @@ private int tryMain(size_t argc, const(char)** argv, ref Param params)
         // If we are here then compilation has ended
         // gracefully as opposed to with `fatal`
         global.plugErrorSinks();
+
+        import core.stdc.stdarg;
+        if (global.errors == 0 && global.params.v.messageStyle == MessageStyle.sarif)
+        {
+            SourceLoc defaultLoc = SourceLoc(null, 0u, 0u);
+            __va_list_tag[1] ap;
```
Comment (thewilsonator) [inline compiler/src/dmd/main.d:181]: `va_list` not `__va_list_tag`

File: compiler/test/compilable/sarif_success_test.d [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+		},
+		"invocations": [{
+			"executionSuccessful": true
+		}],
+		"results": []
+	}]
+}
+---
+*/
+// REQUIRED_ARGS: -verror-style=sarif
```
Comment (thewilsonator) [inline compiler/test/compilable/sarif_success_test.d:23]: this line is usually put at the top before the `TEST_OUTPUT` directive

File: compiler/test/compilable/sarif_success_test.d [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+TEST_OUTPUT:
+---
+{
+	"version": "2.1.0",
+	"$schema": "https://schemastore.azurewebsites.net/schemas/json/sarif-2.1.0.json",
+	"runs": [{
+		"tool": {
+			"driver": {
+				"name": "Digital Mars D",
+				"version": "2.110.0",
```
[the remark this answers is not in the document]
Comment (royalpinto007) [reply compiler/test/compilable/sarif_success_test.d:12]: Do we have any lightweight regex library alternatives to `std.regex` outside Phobos, or should I build a custom parser for version strings (e.g., `v?(\d+(\.\d+)*))`?

[the remark this answers is not in the document]
Comment (royalpinto007) [reply compiler/test/compilable/sarif_success_test.d:12]: something like this- https://github.com/dlang/dmd/pull/17052/commits/d082f9b855fdb56104bcfa3a4e562ff664dee57a

[the remark this answers is not in the document]
Comment (thewilsonator) [reply compiler/test/compilable/sarif_success_test.d:12]: use `$r:<regex>$ ` see https://github.com/dlang/dmd/blob/master/compiler/test/README.md#test_output

[the remark this answers is not in the document]
Comment (royalpinto007) [reply compiler/test/compilable/sarif_success_test.d:12]: Finally, I got it working! 

The issue was with the backslashes. Initially, I used `"$r:\\d+\\.\\d+\\.\\d+$"`, but the double backslashes were treated as literal characters. Changing it to `"$r:\d+\.\d+\.\d+$"` fixed it, as `TEST_OUTPUT` interprets single backslashes correctly in regex patterns

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**